### PR TITLE
Improved way info can be read and written in Virtual Compliance

### PR DESCRIPTION
**Pull request by maxcapodi78**

### CODE REVIEW

File: pyaedt/generic/compliance.py [hunk 1]
```diff
@@ -51,17 +281,102 @@ def __init__(self, desktop, template):
...
         project = self._desktop_class.odesktop.GetActiveProject()
         self._project_name = project.GetName()
         self._output_folder = os.path.join(
             project.GetPath(), self._project_name + ".pyaedt", generate_unique_name(self._template_name)
         )
         os.makedirs(self._output_folder, exist_ok=True)
 
+    @property
+    def reports(self):
+        """Reports available in the Virtual complaince.
```
Comment (MaxJPRey) [inline pyaedt/generic/compliance.py:304]:
```suggestion
        """Reports available in the Virtual compliance.
```

File: pyaedt/generic/compliance.py [hunk 2]
```diff
@@ -30,6 +30,234 @@
...
+        """
+        return self._report_type
+
+    @report_type.setter
+    def report_type(self, val):
+        self._report_type = val
+
+    @property
+    def group_plots(self):
+        """Group plots in single chart or keep independent.
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:69]:
```suggestion
        """Flag indicating if plots are grouped into a single chart or kept independent.
```

File: pyaedt/generic/compliance.py [hunk 3]
```diff
@@ -30,6 +30,234 @@
...
+        """
+        return self._group_plots
+
+    @group_plots.setter
+    def group_plots(self, val):
+        self._group_plots = val
+
+    @property
+    def config_file(self):
+        """configuration file.
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:83]:
```suggestion
        """Configuration file.
```

File: pyaedt/generic/compliance.py [hunk 4]
```diff
@@ -30,6 +30,234 @@
...
+    @traces.setter
+    def traces(self, val):
+        if not isinstance(val, list):
+            self._traces = [val]
+        else:
+            self._traces = val
+
+    @property
+    def pass_fail(self):
+        """Define if apply pass fail criteria.
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:128]:
```suggestion
        """Flag indicating if pass/fail criteria is applied.
```

File: pyaedt/generic/compliance.py [hunk 5]
```diff
@@ -30,6 +30,234 @@
...
+        """
+        return self._trace_pins
+
+    @trace_pins.setter
+    def trace_pins(self, val):
+        self._trace_pins = val
+
+    @property
+    def pass_fail_criteria(self):
+        """Define pass fail criteria.
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:192]:
```suggestion
        """Pass/fail criteria.
```

File: pyaedt/generic/compliance.py [hunk 6]
```diff
@@ -30,6 +30,234 @@
...
+        """
+        return self._pass_fail_criteria
+
+    @pass_fail_criteria.setter
+    def pass_fail_criteria(self, val):
+        self._pass_fail_criteria = val
+
+    @property
+    def config_file(self):
+        """configuration file.
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:206]:
```suggestion
        """Configuration file.
```

File: pyaedt/generic/compliance.py [hunk 7]
```diff
@@ -30,6 +30,234 @@
...
+
+        Returns
+        -------
+        list
+        """
+        self._traces = val
+
+    @property
+    def pass_fail(self):
+        """Define if apply pass fail criteria.
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:248]:
```suggestion
        """Pass/fail criteria.
```

File: pyaedt/generic/compliance.py [hunk 8]
```diff
@@ -39,6 +267,8 @@ class VirtualCompliance:
         Desktop object.
     template : str
         Full path to the template. Supported formats are JSON and TOML.
+    project_path : str, optional
+        Full path to the project to use. If provided, it will ignore the project field inside the template.
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:271]:
```suggestion
    project_path : str, optional
        Full path to the project. If a path is provided, the project field inside the template is ignored.
```

File: pyaedt/generic/compliance.py [hunk 9]
```diff
@@ -51,17 +281,102 @@ def __init__(self, desktop, template):
...
+        self._project_name = None
+        self._output_folder = None
         self._parse_template()
         self._desktop_class = desktop
-        if self._project_file:
-            self._desktop_class.load_project(self._project_file)
+
+    def load_project(self):
+        if not self._project_file:
+            self._desktop_class.logger.error("Project is missed.")
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:292]: Missed? Could the message be clearer?

File: pyaedt/generic/compliance.py [hunk 10]
```diff
@@ -70,24 +385,27 @@ def _parse_template(self):
...
-                    report["config"],
-                    report["design_name"],
-                    report["traces"],
-                    report["pass_fail"],
-                )
+    def _parse_reports(self, report, is_parameter=False):
+        name = report["name"]
+
+        if name in self._reports.values():
+            self._desktop_class.logger.warning(f"{name} already exists. name has to be unique.")
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:403]:
```suggestion
            self._desktop_class.logger.warning(f"{name} already exists. The name must be unique.")
```

File: pyaedt/generic/compliance.py [hunk 11]
```diff
@@ -161,14 +478,17 @@ def _create_aedt_reports(self, pdf_report):
...
-            name = template_report["name"]
-            traces = template_report["traces"]
-            pass_fail = template_report["pass_fail"]
-            design_name = template_report["design_name"]
-            report_type = template_report["type"]
-            group = template_report.get("group_plots", False)
+        for template_report in self._reports.values():
+            config_file = template_report.config_file
+            if not os.path.exists(config_file) and not os.path.exists(os.path.join(self._template_folder, config_file)):
+                self._desktop_class.logger.error(f"{config_file} not found.")
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:484]:
```suggestion
                self._desktop_class.logger.error(f"{config_file} is not found.")
```

File: pyaedt/generic/compliance.py [hunk 12]
```diff
@@ -340,6 +676,10 @@ def _add_lna_violations(self, report, pdf_report, image_name, local_config):
                 "Test Result",
             ]
         ]
+        if not trace_data:  # pragma: no cover
+            msg = "Failed to get Solution Data. Check if the design is solved or the report data are correct."
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:680]:
```suggestion
            msg = "Failed to get solution data. Check if the design is solved or if the report data is correct."
```
US English assumes "data" is a collection of pieces of information and treats it as a singular noun.

File: pyaedt/generic/compliance.py [hunk 13]
```diff
@@ -51,17 +281,102 @@ def __init__(self, desktop, template):
...
         project = self._desktop_class.odesktop.GetActiveProject()
         self._project_name = project.GetName()
         self._output_folder = os.path.join(
             project.GetPath(), self._project_name + ".pyaedt", generate_unique_name(self._template_name)
         )
         os.makedirs(self._output_folder, exist_ok=True)
 
+    @property
+    def reports(self):
+        """Reports available in the Virtual compliance.
```
Comment (PipKat) [inline pyaedt/generic/compliance.py:304]:
```suggestion
        """Reports available in the virtual compliance.
```

File: pyaedt/generic/compliance.py (hunk 9 above)
Comment (maxcapodi78) [reply pyaedt/generic/compliance.py:292]: Sorry, I’ll fix tomorrow

**Review by SMoraisAnsys — APPROVED** [review]
LGTM

**Pull request merged**